java/src/jni/h5dImp.c

```diff
@@ -182,47 +182,69 @@ Java_hdf_hdf5lib_H5_H5Dread(JNIEnv *env, jclass clss, jlong dataset_id, jlong me
                             jlong file_space_id, jlong xfer_plist_id, jbyteArray buf,
                             jboolean isCriticalPinning)
 {
-    jboolean readBufIsCopy;
-    jbyte   *readBuf = NULL;
-    htri_t   vl_data_class;
-    herr_t   status = FAIL;
+    jboolean    readBufIsCopy;
+    jbyte      *readBuf = NULL;
+    size_t      typeSize;
+    H5T_class_t type_class;
+    jsize       n;
+    htri_t      vl_data_class;
+    herr_t      status = FAIL;
 
     UNUSED(clss);
 
     if (NULL == buf)
         H5_NULL_ARGUMENT_ERROR(ENVONLY, "H5Dread: read buffer is NULL");
 
     /* Get size of data array */
-    if (ENVPTR->GetArrayLength(ENVONLY, buf) < 0) {
+    if ((n = ENVPTR->GetArrayLength(ENVONLY, buf)) < 0) {
         CHECK_JNI_EXCEPTION(ENVONLY, JNI_TRUE);
         H5_BAD_ARGUMENT_ERROR(ENVONLY, "H5Dread: readBuf length < 0");
     }
 
     if ((vl_data_class = h5str_detect_vlen(mem_type_id)) < 0)
         H5_LIBRARY_ERROR(ENVONLY);
 
-    if (isCriticalPinning) {
-        PIN_BYTE_ARRAY_CRITICAL(ENVONLY, buf, readBuf, &readBufIsCopy,
-                                "H5Dread: read buffer not critically pinned");
+    if (!(typeSize = H5Tget_size(mem_type_id)))
+        H5_LIBRARY_ERROR(ENVONLY);
+
+    if ((type_class = H5Tget_class((hid_t)mem_type_id)) < 0)
+        H5_LIBRARY_ERROR(ENVONLY);
+    if (vl_data_class) {
+        if (NULL == (readBuf = HDcalloc((size_t)n, typeSize)))
+            H5_OUT_OF_MEMORY_ERROR(ENVONLY, "H5Dread: failed to allocate raw VL read buffer");
     }
     else {
-        PIN_BYTE_ARRAY(ENVONLY, buf, readBuf, &readBufIsCopy, "H5Dread: read buffer not pinned");
+        if (isCriticalPinning) {
+            PIN_BYTE_ARRAY_CRITICAL(ENVONLY, buf, readBuf, &readBufIsCopy,
+                                    "H5Dread: read buffer not critically pinned");
+        }
+        else {
+            PIN_BYTE_ARRAY(ENVONLY, buf, readBuf, &readBufIsCopy, "H5Dread: read buffer not pinned");
+        }
     }
 
     if ((status = H5Dread((hid_t)dataset_id, (hid_t)mem_type_id, (hid_t)mem_space_id, (hid_t)file_space_id,
-                          (hid_t)xfer_plist_id, readBuf)) < 0)
+                          (hid_t)xfer_plist_id, (void *)readBuf)) < 0)
         H5_LIBRARY_ERROR(ENVONLY);
 
+    if (vl_data_class)
+        translate_rbuf(env, buf, mem_type_id, type_class, n, (jobjectArray)readBuf);
```

Comment:
Just a thought - might be worthing naming these `translate_vl_rbuf`/`translate_vl_wbuf` just so there's no confusion at first glance about what type of buffers they're translating.

Reply:
No, because they will also do array and compounds

Reply:
Right now they are only triggered by a vl type, but the function is written such that it isn't a requirement.

Reply:
> Right now they are only triggered by a vl type, but the function is written such that it isn't a requirement.

But isn't that effectively no different than a VL type being a requirement? Also, what if they're triggered by a VL type inside a compound? Doesn't that mean the data for a compound will be in a different format than Java expects since it'll get translated into a list?

Reply:
No, the function just translate raw buffer (bytes) to / from Lists.
So the functions could do just compound or array or any combo of vl, compound or array. The Java buffer is an ArrayList the raw buffer bytes to the hdf5 library